migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/migrate01/MigratorSchema.scala

```diff
@@ -0,0 +1,134 @@
+  override protected val collectionNew: String = MigrationUtils.getVersionedCollectionName(collectionBase, 1)
+
+  protected val dbOld: MongoDatabase = databaseOld
+  protected val dbNew: MongoDatabase = databaseNew
+
+  /**
+    * Migrates a single instance of schema.
+    *
+    * @param srcSchemaJson A JSON representation of a schema in the model 0 format.
+    * @param objectId      An Object Id if the schema.
+    * @param repo          An entity repository.
+    */
+  def migrateEntity(srcSchemaJson: String, objectId: String, repo: EntityRepository): Unit = {
+    val schema1Opt = try {
+      val schema0 = Serializer0.deserializeSchema(srcSchemaJson)
+      Option(model1.Schema(
+        schema0.name,
+        schema0.version,
+        None,
+        userCreated = migrationUserName,
+        userUpdated = migrationUserName,
+        fields = schema0.fields.map(MigrationToV1.convertSchemaField(_, Nil))
+      ))
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Encountered a serialization error for '$collectionBase': ${e.getMessage}")
+        None
+    }
+
+    schema1Opt.foreach(schema1 => {
+      if (repo.doesDocumentExist(schema1.name, schema1.version)) {
+        resolveConflict(schema1, objectId, repo)
+      } else {
+        normalInsert(schema1, objectId, repo)
+      }
+    })
+  }
+
+  /**
+    * In case there is no conflict, insert a new schema normally.
+    * If inserting is unsuccessful fallback to conflict resolution.
+    *
+    * @param schema   A schema to save as an instance of Model 1 object
+    * @param objectId An Object Id of the schema
+    * @param repo     An entity repository
+    */
+  def normalInsert(schema: model1.Schema, objectId: String, repo: EntityRepository): Unit = {
+    val schema1Json = ObjectIdTools.putObjectIdIfNotPresent(Serializer1.serializeSchema(schema), Option(objectId))
+
+    try {
+      repo.insertDocument(schema1Json)
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Unable to append a document for 'schema': ${e.getMessage}. Retrying...")
+        resolveConflict(schema, objectId, repo)
+    }
+  }
+
+  /**
+    * In case there is a conflict, resolve it by adding a new version of the schema
+    * and register the mapping between the old version and the new one in ahe entity version map.
+    *
+    * @param schema   A schema to save as an instance of Model 1 object
+    * @param objectId An Object Id if the schema
+    * @param repo     An entity repository
+    */
+  def resolveConflict(schema: model1.Schema, objectId: String, repo: EntityRepository): Unit = {
+    var retriesLeft = EntityMigrator.NumberOfInsertRetries
+    var saved = false
+    while (retriesLeft > 0 && !saved) {
+      retriesLeft -= 1
+      val latestVersion = repo.getLatestVersion(schema.name)
+      val schemaToSave = schema.copy(version = latestVersion + 1)
+      val schema1Json = ObjectIdTools.putObjectIdIfNotPresent(Serializer1.serializeSchema(schemaToSave),
```

Comment:
can these 3 `val`s be extracted above while?

Reply:
Each time we can potentially have a different value returned from `repo.getLatestVersion(schema.name)`. And that changes values of the next vals.

Reply:
So in case someone created a new version in db_v1 just now and it fails and we need to retry?

Reply:
Exactly